### ISSUE TITLE
Decompilation cannot decompile xUnit's `Assert.All`

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#25246]
Attempted to decompile `Assert.All` from xUnit.Assert (2.1.0) and got the following:

``` C#
		public static void All<T>(IEnumerable<T> collection, Action<T> action)
		{
			//IL_0016: Unknown result type (might be due to invalid IL)
			//IL_001b: Unknown result type (might be due to invalid IL)
			//IL_0038: Unknown result type (might be due to invalid IL)
			//IL_004c: Expected Ref, but got Unknown
			//IL_005d: Unknown result type (might be due to invalid IL)
			//IL_005e: Expected Ref, but got Unknown
			//IL_0069: Unknown result type (might be due to invalid IL)
			//IL_006a: Expected Ref, but got Unknown
			Assert.GuardArgumentNotNull("collection", collection);
			Assert.GuardArgumentNotNull("action", action);
			Stack val = new Stack();
			T[] array = collection.ToArray();
			for (int i = 0; i < array.Length; i++)
			{
				try
				{
					action(array[i]);
				}
				catch (Exception item)
				{
					((Stack)(?)val).Push((!0)new Tuple<int, object, Exception>(i, (object)array[i], item));
				}
			}
			if (((Stack)(?)val).get_Count() <= 0)
			{
				return;
			}
			throw new AllException(array.Length, (Tuple<int, object, Exception>[])((Stack)(?)val).ToArray());
		}
```

**Comment by davkean** [comment]
Actually there's a bunch in that class that couldn't be decompiled:
```
Assert.PropertyChanged
Assert.PropertyChangedAsync
Assert.ProperSubset<T>
Assert.ProperSuperset<T>
Assert.Superset<T>
Assert.Matches
```

**Comment by davkean** [comment]
This is probably related to: https://github.com/dotnet/roslyn/issues/25310.

**Comment by sharwell** [comment]
This is caused by failure to resolve a dependency of xunit during decompilation. It's the same type of problem that led to #24178, but involves a different dependency. Once we are able to correctly identify and resolve the dependency, several decompilation issues in this file will be fixed. I'm going to mark those issues as a duplicate of this one.

**Comment by CyrusNajmabadi** [comment]
> This is caused by failure to resolve a dependency of xunit during decompilation. It's the same type of problem that led to #24178, but involves a different dependency. Once we are able to correctly identify and resolve the dependency, several decompilation issues in this file will be fixed. I'm going to mark those issues as a duplicate of this one.

That seems like the best course of action.  I do wonder though if its worth investing a tiny amount of effort in massaging the text that ilspy returns before using it.  Say, for example, a dependency truly is missing.  If ilspy then generates things like ```(?)``` and ```(!0)``` then that's always going to be an issue for us.  We could consider doing a very simple regex find/replace of known bad patterns to get the code back into a reasonable state so that things parse again.

Or, we can ask ilspy if they can change how they decompile in these situations (perhaps gated by a flag).  i.e. do not spit out casts to unknown types such that bogus code would be produced.  

If that info is considered important, perhaps it could be emitted as ```.Push(/*(!0)*/new Tuple``` instead of ```.Push((!0)new Tuple```.

**Comment by dgrunwald** [comment]
Since VS is re-parsing the decompiled code, it should probably do:
```
decompiler.AstTransforms.Add(new EscapeInvalidIdentifiers());
```
This makes the code look even worse (invalid characters are replaced an underscore and hex character code), but at least it should stop the parser from getting confused.
I'll take a look at modifying the decompiler to suppress casts when we don't have a valid name for the target type.

**Comment by CyrusNajmabadi** [comment]
@dgrunwald Thanks!  

> Since VS is re-parsing the decompiled code

Yup, and that's a core part of the design, as that allows to enable all the other sorts of features that users expect with source code (i.e. nav-bars, outlining, etc.).  So it's really important that we have a good tree to work with even in problematic scenarios.

>I'll take a look at modifying the decompiler to suppress casts when we don't have a valid name for the target type.

Thanks.  That would be great!

**Comment by dgrunwald** [comment]
I did some improvements to how we represent unresolved types in the ILSpy type system (especially in their interaction with generics), so `Assert.All` now decompiles almost correctly when the reference assemblies are missing. This will be part of the upcoming ILSpy 3.1 release.
Unresolved reference types should work mostly fine now, but unresolved value types are still a big problem: without having the defining assembly loaded, we can't tell whether a value type is a struct or enum, and that has significant effects on the decompiler.

Visual Studio still needs to:
 * enable the `EscapeInvalidIdentifiers` transform, so that problems with decompilation don't mess up the formatting and other IDE features
 * fix loading of the reference assemblies
    * ILSpy requires that direct references of the assembly being decompiled are available.
    * Additionally, if there are any type forwarders in the direct references, the assemblies actually defining those types need to be available.
    * Reference assemblies are sufficient; only the assembly being decompiled has to contain actual IL code.

**Comment by siegfriedpammer** [comment]
> fix loading of the reference assemblies
> * ILSpy requires that direct references of the assembly being decompiled are available.
> * Additionally, if there are any type forwarders in the direct references, the assemblies actually defining those types need to be available.
> * Reference assemblies are sufficient; only the assembly being decompiled has to contain actual IL code.

@sharwell how are we going to address these? If you need any more info, do not hesitate to ask!

**Comment by davkean** [comment]
>ILSpy requires that direct references of the assembly being decompiled are available.

Why is that? Other than the size of enums in serialized attributes, an assembly contains all the information you need to decompile?

**Comment by dgrunwald** [comment]
> Other than the size of enums in serialized attributes, an assembly contains all the information you need to decompile?

Nope, there's lots missing. Here's an incomplete list:

* We don't know whether enum types are 32-bit or 64-bit, which influences more than just attributes:
    * the implicit width of IL instructions like `add`
    * static array initializer decoding
* We don't know the names for enum constants, so all your `StringComparison.Ordinal` just turned into `(StringComparison)5`.
* We don't know whether static methods being called are extension methods.
* We don't for sure whether a `get_X()` method being called belongs to a property (though we could guess based on the typical naming convention).
* We don't know other overloads similar to the one being called. We would have to conservatively cast every argument, and always use explicit type parameters, just to be sure our generated C# ends up calling the correct overload.
* We don't know whether there are other types in the imported namespaces that collide with the one we'd want to use. We'd have to always use fully qualified type names,
* We don't know whether a type implements IEnumerable, so we can't be sure if we may use collection initializer syntax.

So code like `new List<string> { "Test" }.Where(t => x.StartsWith(t, StringComparison.Ordinal))` could in the best case only decompile to this monster:
```
var tmp = new System.Collections.Generic.List<string>();
tmp.Add("Test");
System.Linq.Enumerable.Where<string>((System.Collections.Generic.IEnumerable<string>)tmp, (System.Func<string, bool>)(t => x.StartsWith(t, (StringComparison)5);
```

I say "best case" because this still requires ILSpy to deal with unknown types in plenty of transforms; e.g. we would have to generate the lambda syntax despite not knowing that `System.Func` is a delegate type or what its signature is (though we could guess based on the compiler-generated method's signature).
